File: FastMines_core/src/fmg/core/app/serializers/ChampionsSerializer.java
package fmg.core.app.serializers;

import java.io.*;
import java.util.Date;
import java.util.List;
import java.util.UUID;

import fmg.common.Logger;
import fmg.core.app.model.Champions;
import fmg.core.app.model.Champions.Record;
import fmg.core.types.EMosaic;
import fmg.core.types.ESkillLevel;

/** Champions base (de)serializer */
public abstract class ChampionsSerializer implements ISerializer {

    protected static final long VERSION = 1;

    private static class RecordSerializer implements ISerializer {

        void write(Champions.Record rec, ObjectOutput to) throws IOException {
            to.writeUTF(rec.userId.toString());
            to.writeUTF(rec.userName);
            to.writeLong(rec.playTime);
            to.writeInt(rec.clicks);
            to.writeLong(rec.date.getTime());
        }

        Champions.Record read(ObjectInput from) throws IOException {
            Champions.Record res = new Record();
            res.userId = UUID.fromString(from.readUTF());
            res.userName = from.readUTF();
            res.playTime = from.readLong();
            res.clicks   = from.readInt();
            res.date = new Date(from.readLong());
            return res;
        }
    }

    private void write(Champions champions, ObjectOutput to) throws IOException {
        to.writeLong(VERSION);
        List<Record>[][] all = champions.getRecords();
        RecordSerializer rs = new RecordSerializer();
        for (EMosaic mosaic : EMosaic.values())
            for (ESkillLevel eSkill : ESkillLevel.values())
                if (eSkill != ESkillLevel.eCustom) {
                    List<Champions.Record> list = all[mosaic.ordinal()][eSkill.ordinal()];
                    to.writeInt(list.size());
                    for (Record record : list)
                        rs.write(record, to);
                }
    }

    private Champions read(ObjectInput from) throws IOException {
        long version = from.readLong();
        if (version != VERSION)
            throw new RuntimeException("Unsupported " + Champions.class.getSimpleName() + " version " + version);

        Champions res = new Champions();
        List<Record>[][] all = res.getRecords();
        RecordSerializer rs = new RecordSerializer();
        for (EMosaic mosaic : EMosaic.values())
            for (ESkillLevel eSkill : ESkillLevel.values())
                if (eSkill != ESkillLevel.eCustom) {
                    List<Champions.Record> list = all[mosaic.ordinal()][eSkill.ordinal()];
                    int size = from.readInt();
                    for (int i=0; i<size; i++) {
                        Record record = rs.read(from);
                        list.add(record);
                    }
                }
        return res;
    }

    /** serialize to bytes */
    private byte[] asBytes(Champions champions) throws IOException {
        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
             ObjectOutputStream oos = new ObjectOutputStream(baos))
        {
            write(champions, oos);
            oos.flush();

            return baos.toByteArray();
        }
    }

    /** deserilize from bytes */
    protected Champions fromBytes(byte[] data) throws IOException {
        try (ByteArrayInputStream bais = new ByteArrayInputStream(data);
             ObjectInputStream ois = new ObjectInputStream(bais))
        {
            return read(ois);
        }
    }

    /** write to file */
    private void write(byte[] data, File file) throws IOException {
        try (OutputStream fos = new FileOutputStream(file);
             ObjectOutputStream oot = new ObjectOutputStream(fos))
        {
            oot.writeLong(VERSION); // save version and decrypt key
            int len = data.length;
            oot.writeInt(len);
            oot.write(data);
        }
    }

    /** read from file */
    private byte[] read(File file) throws IOException {
        try (InputStream fis = new FileInputStream(file);
             ObjectInputStream oin = new ObjectInputStream(fis))
        {
            long version = oin.readLong();
            if (version != VERSION)
                throw new IOException("Invalid file data. Unsupported " + Champions.class.getSimpleName() + " version " + version);

            byte[] data = new byte[oin.readInt()];
            int read = 0;
            do {
                int curr = oin.read(data, read, data.length-read);
                if (curr < 0)
                    break;
                read += curr;
            } while(read < data.length);

            if (read != data.length)
                throw new IOException("Invalid data length. Required " + data.length + " bytes; read " + read + " bytes.");

            return data;
        }
    }

    protected byte[] writeTransform(byte[] data) throws IOException {
        // defaut none
        return data;
    }

    protected byte[] readTransform(byte[] data) throws IOException {
        // defaut none
        return data;
    }

    public void save(Champions champions) {
        try {
            // 1. serialize
            byte[] data = asBytes(champions);

            // 2. transform data
            data = writeTransform(data);

            // 3. write to file
            write(data, getChampionsFile());

        } catch (Exception ex) {
            Logger.error("Can`t save " + Champions.class.getSimpleName(), ex);
        }
    }

    public Champions load() {
        File file = getChampionsFile();
        if (!file.exists())
            return new Champions();

        try {
            // 1. read from file
            byte[] data = read(file);

            // 2. transform data
            data = readTransform(data);

            // 3. deserialize
            return fromBytes(data);

        } catch (Exception ex) {
            Logger.error("Can`t load " + Champions.class.getSimpleName(), ex);
            return new Champions();
        }
    }

    protected abstract File getChampionsFile();

}

File: FastMines_core/src/fmg/core/app/serializers/ChampionsSerializer.cs
using System;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using System.Collections.Generic;
using Fmg.Common;
using Fmg.Core.Types;
using Fmg.Core.App.Model;

namespace Fmg.Core.App.Serializers {

    /// <summary> Champions base (de)serializer </summary>
    public abstract class ChampionsSerializer : ISerializer {

        protected const long VERSION = 1;

        class RecordSerializer : ISerializer {

            public void Write(Champions.Record rec, BinaryWriter output) {
                output.Write(rec.userId.ToString());
                output.Write(rec.userName);
                output.Write(rec.playTime);
                output.Write(rec.clicks);
                output.Write(rec.date.Ticks);
            }

            public Champions.Record Read(BinaryReader input) {
                Champions.Record res = new Champions.Record();
                res.userId   = new Guid(input.ReadString());
                res.userName = input.ReadString();
                res.playTime = input.ReadInt64();
                res.clicks   = input.ReadInt32();
                res.date     = new DateTime(input.ReadInt64());
                return res;
            }

        }

        private void Write(Champions champions, BinaryWriter to) {
            to.Write(VERSION);
            List<Champions.Record>[,] all = champions.Records;
            RecordSerializer rs = new RecordSerializer();
            foreach (var mosaic in EMosaicEx.GetValues())
                foreach (var eSkill in ESkillLevelEx.GetValues())
                    if (eSkill != ESkillLevel.eCustom) {
                        IList<Champions.Record> list = all[mosaic.Ordinal(), eSkill.Ordinal()];
                        to.Write(list.Count);
                        foreach (Champions.Record record in list)
                            rs.Write(record, to);
                    }
        }

        private Champions Read(BinaryReader from) {
            long version = from.ReadInt64();
            if (version != VERSION)
                throw new ArgumentException("Unsupported " + nameof(Champions) + " version " + version);

            Champions res = new Champions();
            List<Champions.Record>[,] all = res.Records;
            RecordSerializer rs = new RecordSerializer();
            foreach (var mosaic in EMosaicEx.GetValues())
                foreach (var eSkill in ESkillLevelEx.GetValues())
                    if (eSkill != ESkillLevel.eCustom) {
                        IList<Champions.Record> list = all[mosaic.Ordinal(), eSkill.Ordinal()];
                        int size = from.ReadInt32();
                        for (int i=0; i<size; i++) {
                            Champions.Record record = rs.Read(from);
                            list.Add(record);
                        }
                    }
            return res;
        }

        /// <summary> serialize to bytes </summary>
        private async Task<byte[]> AsBytes(Champions champions) {
            using (Stream stream = new MemoryStream()) {
            using (BinaryWriter output = new BinaryWriter(stream, Encoding.UTF8)) {
                Write(champions, output);
                output.Flush();

                byte[] data = new byte[stream.Length];
                stream.Position = 0;
                var readed = await stream.ReadAsync(data, 0, data.Length);
                if (readed != data.Length)
                    throw new Exception("Not readed saved data: requred " + data.Length + " bytes; read " + readed + " bytes");
                return data;
            }}
        }

        /// <summary> deserilize from bytes </summary>
        private async Task<Champions> FromBytes(byte[] data) {
            using (Stream stream = new MemoryStream()) {
                await stream.WriteAsync(data, 0, data.Length);
                stream.Position = 0;
                using (BinaryReader input = new BinaryReader(stream)) {
                    return Read(input);
                }
            }
        }

        /// <summary> write to file </summary>
        protected abstract Task Write(byte[] data, string file);

        /// <summary> read from file </summary>
        protected abstract Task<byte[]> Read(string file);

        protected virtual byte[] WriteTransform(byte[] data) {
            // defaut none
            return data;
        }

        protected virtual byte[] ReadTransform(byte[] data) {
            // defaut none
            return data;
        }

        public async Task Save(Champions champions) {
            Logger.Debug("> ChampionsSerializer::Save");
            try {
                // 1. serialize
                byte[] data = await AsBytes(champions);

                // 2. transform data
                data = WriteTransform(data);

                // 3. write to file
                await Write(data, GetChampionsFile());

            } catch (Exception ex) {
                Logger.Error("Can`t save " + nameof(Champions), ex);
            } finally {
                Logger.Debug("< ChampionsSerializer::Save");
            }
        }

        public async Task<Champions> Load() {
            Logger.Debug("> ChampionsSerializer::Load");
            try {
                string file = GetChampionsFile();
                if (!await IsFileExist(file))
                    return new Champions();

                // 1. read from file
                byte[] data = await Read(file);

                // 2. transform data
                data = ReadTransform(data);

                // 3. deserialize
                return await FromBytes(data);

            } catch (Exception ex) {
                Logger.Error("Can`t load " + nameof(Champions), ex);
                return new Champions();
            } finally {
                Logger.Debug("< ChampionsSerializer::Load");
            }
        }

        protected abstract string GetChampionsFile();
        protected abstract Task<bool> IsFileExist(string file);

    }

}
